assert/assertions.go

```diff
@@ -598,6 +598,56 @@ func NotEqual(t TestingT, expected, actual interface{}, msgAndArgs ...interface{
 
 }
 
+// assertIncludeElementTypes checks whether types are compatible as incompatible types indicate programmer error.
+// return true if list and element types are compatible and includeElement can possibly return found=true.
+// return false if list and element types aren't compatible and includeElement will always return found=false.
+func assertIncludeElementTypes(list interface{}, element interface{}, elementIsList bool) bool {
+
+	var elementType reflect.Type
+	if element != nil {
```

Comment:
We can't blindly "unwrap" `element` if it is of kind `reflect.Array`, because it'll break such constructs as `Contains(t, [...][1]int{{1}, {2}, {3}}, [1]int{1})`. "Unwrapping" `reflect.Slice` is also dangerous: `Contains(t, [][]int{{1}, {2}, {3}, nil}, []int(nil))`. It is possible to unwrap `element` with respect to `list` type. But it'll increase complexity of the code without much profit.
Another problem is that it will break things like `Subset(t, []int{1, 2}, int(0))`. Impossible type combination but `assertIncludeElementTypes` will be happy.
Here is implementation of `assertIncludeElementTypes` without `elementIsList`, for the reference:
```
func assertIncludeElementTypes(list interface{}, element interface{}) bool {
	var listElementType reflect.Type
	switch listType := reflect.TypeOf(list); listType.Kind() {
	case reflect.String:
		listElementType = listType

	case reflect.Map:
		listElementType = listType.Key()

	case reflect.Slice, reflect.Array:
		listElementType = listType.Elem()

	default:
		return false
	}

	if element == nil {
		switch listElementType.Kind() {
		case reflect.Ptr, reflect.Slice, reflect.Map, reflect.Interface, reflect.Func:
			return true
		default:
			return false
		}
	}

	isCompatible := func(t reflect.Type) bool {
		if t.Kind() == reflect.Interface {
			// Anything can be hidden behind interface
			return true
		}

		return t.AssignableTo(listElementType)
	}

	var elementType reflect.Type
	elementType = reflect.TypeOf(element)
	if !isCompatible(elementType) {
		switch elementType.Kind() {
		case reflect.Slice, reflect.Array:
			if !isCompatible(elementType.Elem()) {
				return false
			}

			elementType = elementType.Elem()

		default:
			return false
		}
	}

	return true
}
```
It is broken in ways I've described above.